### PR TITLE
AI Tutor / AI Chat - Separate API keys by client and environment

### PR DESCRIPTION
**Pull request by edcodedotorg**
This PR replaces the general `CDO.google_gemini_student_learning_api_key` (which was the same key for development, test, and production) with client (AI Tutor / AI Chat) specific `CDO.google_gemini_ai_chat_api_key` and `CDO.google_gemini_ai_tutor_api_key` keys backed  by environment specific (development / test / production) keys in GCP (Google Cloud Platform) [cloud console](https://console.cloud.google.com/apis/credentials) and AWS [Secrets Manager](https://us-east-1.console.aws.amazon.com/secretsmanager).

## Links
The Jira ticket is [here](https://codedotorg.atlassian.net/browse/LABS-1556?atlOrigin=eyJpIjoiM2VlYmZlZTgwN2E5NGRmMjgzYmM4NjY2ZjEwODVhYmMiLCJwIjoiaiJ9) and this has been discussed in various slack threads including this [one](https://codedotorg.slack.com/archives/C06FELVTV0Q/p1758315431672639).  

## Testing story
The change to how `AichatAiClient` instances are created (moved factory method to helper), required an update to `aichat_ai_client_test.rb`, otherwise tests are not affected.

## Follow-up work
The old key (CDO.google_gemini_student_learning_api_key) should be removed from AWS Secrets Manager.

This PR addresses the two main AI clients (AI Tutor and Chat) but does not address Flow Lab. As it seems the `FLOW_LAB` client id is used for experimentation, perhaps we should rename it as EXPERIMENTS or EXPERIMENTAL and add additional keys (per env) for use in experiments.  Currently flow lab and any experiments identifying as flow lab are using the AI Chat key.

## PR Creation Checklist:
- [ ] Tests provide adequate coverage
- [ ] Privacy impacts have been documented
- [ ] Security impacts have been documented
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by bencodeorg** [comment]
Could you say more about having per-client keys? Looking at the thread, sounds like it might have something to do with donated funds? Or is it just a code organization decision?

**Comment by edcodedotorg** [comment]
> Could you say more about having per-client keys? Looking at the thread, sounds like it might have something to do with donated funds? Or is it just a code organization decision?

Will defer to @cat5inthecradle for the official answer.  Will message you with what I know offline.

File: dashboard/app/helpers/aichat_ai_helper.rb [hunk 1]
```diff
@@ -153,6 +153,19 @@ def self.get_config_request_context(stored_messages, new_message, temperature, s
...
 
+  # Create an instance of the appropriate ai-client-derived class based on model id.
+  def self.create_ai_client_instance(client_type, model_id, usage_reporter = nil)
+    # We assume it's one of the gemini models if not 'gpt-4o-mini'.
+    if model_id == "gpt-4o-mini"
+      return AichatOpenaiResponsesClient.new(CDO.openai_student_learning_api_key, SharedConstants::AICHAT_MODEL_VERSION, usage_reporter)
+    else
+      # We assume AI Chat if not tutor - todo: rename FLOW_LAB to EXPERIMENTS and create a key for it.
+      api_key = client_type == SharedConstants::AI_CHAT_CLIENT_TYPES[:AI_TUTOR] ?
+        CDO.google_gemini_ai_tutor_api_key : CDO.google_gemini_ai_chat_api_key
```
Comment (bencodeorg) [inline dashboard/app/helpers/aichat_ai_helper.rb]: Nit: could we name this `google_gemini_ai_chat_lab_api_key` (ie, with lab in the name)? It's not the most clear, but at least gives us some differentiation when referring to the frontend AI Chat Lab client vs. the back end AI Chat system that serves multiple clients. Same with the comment on line 162.

Comment (bencodeorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: Also a note about why we have client-specific API keys for Gemini, but not for OpenAI?

File: dashboard/app/helpers/aichat_ai_helper.rb [hunk 2]
```diff
@@ -153,6 +153,19 @@ def self.get_config_request_context(stored_messages, new_message, temperature, s
     return config, request, context
   end
 
+  # Create an instance of the appropriate ai-client-derived class based on model id.
```
Comment (bencodeorg) [inline dashboard/app/helpers/aichat_ai_helper.rb:156]: For my understanding, is the move to create the instances in this file instead of in `aichat_ai_client` necessary for this change, or for tidiniess?

**Review by bencodeorg — APPROVED** [review]
LGTM (assuming the keys have been / will be created before merging?)

**Comment by cat5inthecradle** [comment]
> Could you say more about having per-client keys? Looking at the thread, sounds like it might have something to do with donated funds? Or is it just a code organization decision?

Both. Separate keys (specifically, separate GCP projects) allow to track and fund usage separately. We can monitor and spot problems and easily identify which feature and environment is seeing abnormal usage.

File: dashboard/app/helpers/aichat_ai_helper.rb (hunk 2 above)
Comment (edcodedotorg) [reply dashboard/app/helpers/aichat_ai_helper.rb:156]: _TLDR:_ I wouldn't say it was strictly necessary, but yes it was for tidiness, but no, it wasn't something that needed cleaning up before this change (it's not what we've been calling an opportunistic refactor). 

_Longer answer:_ It's mostly about separation of concerns and emerging clarity around use cases.  The goal is for the `ai_client` classes to be use-case agnostic and this change requires us to select the API key based on `client_id` which is AI Chat / Tutor specific (other uses won't have a concept of a `client_id` and they may have different keys based on our key per use-case policy). So this concern doesn't belong in what should be an agnostic `aichat_ai_client.rb`.  

But this selection of API key seemed to make sense in a helper (e.g. instances need to be created in the unit tests and all instances need a key even if we're stubbing the actual http request - I didn't want to duplicate or recreate the key selection logic outside of a helper that could be shared) so we'd either need two helpers (original `create_instance()` and a new `get_api_key()`) with the former in `aichat_ai_client.rb` and the latter in `aichat_ai_helper.rb`, or consolidate in one helper (the new `create_ai_client_instance()` you referenced) just in `aichat_ai_helper.rb`.  

When thinking about the next planned use of the client, it was clear that the dynamic selection of AI provider/backend is not universal (some use cases will just pick Gemini or OpenAI) so moving the selection of backend outside of `aichat_ai_client.rb` seemed like the right call given the new requirements (key based on `client_id`) and clearer overall picture.

File: dashboard/app/helpers/aichat_ai_helper.rb (hunk 1 above)
Comment (edcodedotorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: Ok, the `_lab` addition makes sense.  But you said it's not the most clear.  Do you have a completely different alternative that is clearer?  If we're changing, it doesn't have to be a small change ;)

Not sure if we have separate environment specific keys for OpenAI. If not, that should be possible without a code change.  I'll leave the question of whether we add client specific keys for OpenAI to the PMs and infra, but we can spin it up if/when prioritized.

**Comment by edcodedotorg** [comment]
> LGTM (assuming the keys have been / will be created before merging?)

thanks and yes, waiting on infra to settle on the key distribution which is being discussed [here](https://codedotorg.slack.com/archives/C03CK49G9/p1758643419538119).

File: dashboard/app/helpers/aichat_ai_helper.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: I would roll with the `_lab` version for now. We've been trying to separate references to "AI Chat Lab" and "AI Chat backend" (eg, in the name of the clients):

https://github.com/code-dot-org/code-dot-org/blob/88e7a33359e6a1782c24578dc10e5e64c38b1d3f/lib/cdo/shared_constants.rb#L834-L838

and I think that key name would line up with that convention (still confusing though, as you noted). Maybe longer term we work toward a rename of the backend code?

Re: OpenAI, I just meant adding a comment noting why we have client-specific (not env-specific) keys for Gemini, but not for OpenAI.

**Comment by bencodeorg** [comment]
Re: the per-client / more keys question in general, I was nudged towards [fewer keys for OpenAI earlier this year](https://codedotorg.slack.com/archives/C06FELVTV0Q/p1746566530469009?thread_ts=1739319271.612899&cid=C06FELVTV0Q), but kind of regret it.

Our logging in CloudWatch doesn't quite line up with what we see in OpenAI's dashboards, and using a single key/project made answering simple questions via the OpenAI interface (ie, how much are we spending on this feature / project?) harder than expected. The tradeoff of managing more keys feels worthwhile IMO.

File: dashboard/app/helpers/aichat_ai_helper.rb (hunk 1 above)
Comment (edcodedotorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: oh haha I read `note` as your comment was a note.  I don't have a specific reason other than this is what was prioritized.

**Comment by cat5inthecradle** [comment]
+1 to Ben's comment. We should still strive for quality metrics within our platform, but by using distinct API keys, we get a second set of data via the vendor's metrics which we can correlate.

Edit-to-add: An important factor is "How easy is it to make more keys", and in the case of Gemini API keys, it's pretty easy, the tedium is that I'm simultaneously setting up some org-wide governance and organization, which should make future GCP work more straightforward.

File: dashboard/app/helpers/aichat_ai_helper.rb [hunk 3]
```diff
@@ -153,6 +153,19 @@ def self.get_config_request_context(stored_messages, new_message, temperature, s
...
+  def self.create_ai_client_instance(client_type, model_id, usage_reporter = nil)
+    # We assume it's one of the gemini models if not 'gpt-4o-mini'.
+    if model_id == "gpt-4o-mini"
+      return AichatOpenaiResponsesClient.new(CDO.openai_student_learning_api_key, SharedConstants::AICHAT_MODEL_VERSION, usage_reporter)
+    else
+      # We assume AI Chat if not tutor - todo: rename FLOW_LAB to EXPERIMENTS and create a key for it.
```
Comment (bencodeorg) [inline dashboard/app/helpers/aichat_ai_helper.rb]:
```suggestion
      # We use separate keys per-client for Gemini so that we can more easily allocate donated credits appropriately.
      # In the longer term, we should consider adding per-client keys for OpenAI as well in order to more easily differentiate between use cases.
      # Also note that we assume AI Chat Lab if not AI Tutor here, which is not strictly true because of Flow Lab. We could add an EXPERIMENT client type and add an explicit key to handle those use cases (eg, Flow Lab). 
```

File: dashboard/app/helpers/aichat_ai_helper.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: Tried out GitHub's suggestion feature (not sure if I used it right 😅 ) with a comment along the lines of what I was thinking.

Comment (edcodedotorg) [reply dashboard/app/helpers/aichat_ai_helper.rb]: Accepted thanks
